userchroot.c

```diff
@@ -227,96 +223,128 @@ static void portable_clearenv() {
-    d->argv++;d->argv++;
-    whitelist_char_check(d->argv[0], 1);
-    execve(d->argv[0],d->argv,d->envp);
-    // if we are here, it means something went wrong.
-    fprintf(stderr,"Failed to exec %s: %s\n", d->argv[0], strerror(errno));
+  rc = chdir("/");
+  if (rc != 0) {
+    fprintf(stderr,"Failed to chdir to the root directory. Aborting.\n");
     exit(ERR_EXIT_CODE);
+  }
+
+  // And finally, execute the desired command.
+  // we skip the first two arguments from argv and do a execve.
+  d->argv++;d->argv++;
+  whitelist_char_check(d->argv[0], 1);
+  execve(d->argv[0],d->argv,d->envp);
+  // if we are here, it means something went wrong.
+  fprintf(stderr,"Failed to exec %s: %s\n", d->argv[0], strerror(errno));
+  exit(ERR_EXIT_CODE);
 }
 
-#ifdef __linux__
-  #ifdef MOUNT_PROC
-    #if LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,24)
+#if defined (__linux__) && defined (MOUNT_PROC)
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,24))
 
 static char child_stack[1048576];
+static char proc_guard_stack[1048576];
 
 static int child_fn(void* v) {
-
   struct epilogue_data* ed = (struct epilogue_data*)v;
+  epilogue(ed);
+  return 0;
+}
+
+static int proc_guard(void *v) {
 
-  //Since we're in the chroot, we don't need to unmount the current
-  //proc, simply because there isn't any current proc mounted.
+  // Since we're in the chroot, we don't need to unmount the current
+  // proc, simply because there isn't any current proc mounted.
   //
-  //Mount the child's view of proc, which includes only processes
-  //in its namespace
+  // Mount the child's view of proc, which includes only processes
+  // in its namespace.
   int rc = mkdir("/proc", S_IRWXU);
   if(0 != rc && EEXIST != errno) {
     fprintf(
-      stderr,
-      "Failed to mkdir /proc. Error: %s\n", strerror(errno)
-    );
+            stderr,
+            "Failed to mkdir /proc. Error: %s\n", strerror(errno)
+            );
+    return rc;
   }
   else {
     rc = mount(
-            "proc",
-            "/proc",
-            "proc",
-            MS_REC|MS_NOSUID|MS_NODEV|MS_NOEXEC,
-            NULL
-         );
+               "proc",
+               "/proc",
+               "proc",
+               MS_REC|MS_NOSUID|MS_NODEV|MS_NOEXEC,
+               NULL
+               );
     if(0 != rc) {
       fprintf(
-        stderr,
-        "Failed to mount proc. Error: %s\n", strerror(errno)
-      );
+              stderr,
+              "Failed to mount proc. Error: %s\n", strerror(errno)
+              );
+      return rc;
     }
   }
 
-  epilogue(ed);
-  return 0;
-}
+  pid_t child_pid =
+    clone(
+          child_fn,
+          child_stack+sizeof(child_stack),
+          SIGCHLD,
+          v
+          );
+  if(-1 == child_pid) {
+    fprintf(stderr, "Failed to clone. Error: %s\n", strerror(errno));
+  }
+  else {
+    int child_status = 0;
+    int p = 0;
+    while (p = waitpid(child_pid, &child_status, 0)) {
+      if (p == child_pid || p == -1) {
```

Comment:
if p == -1, child_status will not be set (I'm pretty sure), and in the end it will exit with status 0. Maybe the default initialization of child_status needs to be something else...